### ISSUE TITLE
Improve: config settings in the annotator 

### ISSUE THREAD

**Issue by LMS007** [hypothesis/client#3236]
## 1. Current implementation issues

Currently, the configuration settings in the annotator are gathered from `settings.js` and pulled into an object in `config/index.js`. There are several issues with the current implementation that are confusing to read and could be slightly improved

### 1a. Default Value

All values are set to a default value of null if they are missing / undefined. But there is also a `defaultValue` setting that can further override this null value.  Can we simply only use `defaultValue` and anything not defined can be just omitted from the config? At the very least, don't bake in null values and defined a `defaultValue` set to null for each item that needs to be null. At least this way its explicit from looking at a configuration definition. 

### 1b. Coerce methods

Currently, only one value has a coerce method (several more in the sidebar--unrelated to this ticket). We do need to maintain a way of augmenting values because incoming config may have older standards we need to maintain for backwards compatibility. Secondly, we need to support the more limited types that via can produce. e.g. It only supports strings. 

`showhighlights` also needs the be changed (depending on logic) and its currently done in `settings.js`. It seems that much of that logic is suitable for a coerce method similar to `openSidebar`. It's confusing that logic to coerce values is in multiple places.

### 1c. Browser extension context

All values are subject to being removed/nulled if in the browser extension context. This is separate from an application context because both can be true or worth orthogonally. In other words, we still need to honor `allowInBrowserExt` in the notebook context or sidebar, etc. It makes sense that each setting gets an `allowInBrowserExt` flag. Currently, this is not the case. All settings that come from `hostPageSetting` do honor the `allowInBrowserExt` flag (defaults to `false`), but the getter methods in `settings.js` no not honor the flag. In fact, `showHighlights` even calls down into `hostPageSetting()` which has the ability to check the flag, but `showHighlights` does not actually pass an options object down so it misses the opportunity. While it may be true that all the config settings derived from such getters should be allowed in the browser extension context, it's unclear from the code that they should be allowed in the first place. In other words, it should be expressed clearly that they are allowed or not.

getters:

```
annotations()
clientUrl()
group()
notebookAppUrl()
showHighlights()
sidebarAppUrl() 
query()
```

## 2. Resulting configs need to be different distinct applications.  (New feature)

Currently, we only have 1 config, but now that we have a notebook and an actual need to mutate the config before it gets passed to the notebook, it makes sense for the thing that "gets" the config to be able to filter to specific application contexts. For for example, the notebook does not get `annotations` setting because it may affect the rendering threads to how they should be rendered in the notebook context. Furthermore, an audit of the config vars may in fact reveal that the sidebar and annotator do not necessarily need the entire configuration either and it may improve readability to only pass along key/values which are indeed used.

## 3. Valitation (New feature)

At some point, we may need to implement settings validation so that variables coming from the various config entry points first pass through a sanity check. At this point, we could inform the user of a warning or error and explain which values are rejected or are erroneous, etc.

For example
```
theme: 'fun',
fakeValue: true,
```

The first is an invalid value, the second is an invalid key. We should probably inform the user (presumably an admin embedding the client) rather than silently nulling or omitting things.

## 4. Define an API that annotator/index.js can use.

Proposed API (though others could work just as well)

`annotator/index.js`

```
import { configFor } from 'config/index'

var configA = configFor('annotator')
var configS= configFor('sidebar')
var configN = configFor('notebook')
```

**A pipeline may work like this:**

1. Gather the values from respective sources (json, config function, url, etc)
2. Validate the values against defined rules for each key
3. Browser extension check: omit keys if they fail the check 
4. Further omit out keys that are out of scope of a requested app context. 
5. For remaining keys, if corresponding values are undefined and a default exists, set the default, otherwise, run them through coerce() and return the final (app scoped) config.

### In summary there are 6 parts to this ticket.

- [x] refactored config may need a new API that operates off a unique app name
- [x] configs should have an explicit default when the key is missing, omit otherwise
- [x] configs should have an explicit flag that allows them in the browser extension, and omit otherwise 
- [x] configs should have a single place to coerce the values is necessary .e.g "true" to `true`
- [x] configs should explicitly have an an allowed applications ('notebook', 'sidebar', 'annotator') context, and omit otherwise.
- [ ] configs needs a single place (for each iframe)  where the config passes / fails a validation type/function. Invalid configs keys/values needs to be surfaced to the web admin (perhaps in the console log) and omitted thereafter, (note, don't set to default it value fails validation, that may be confusing)

**Comment by robertknight** [comment]
There is an update on the overall status of this at https://github.com/hypothesis/client/pull/3501#issuecomment-863218121. The TL;DR is that the current set of open PRs only needs some minor changes before we can merge them. Once those are deployed, I think we will have reached a point where it makes sense to consider this issue resolved.

**Comment by LMS007** [comment]
The validation will not be done as part of this ticket, all other requirements have been met.
